python/pylibcugraph/pylibcugraph/node2vec.pyx

```diff
@@ -0,0 +1,184 @@
+    >>> weights = cupy.asarray([1.0, 1.0, 1.0], dtype=numpy.float32)
+    >>> resource_handle = pylibcugraph.experimental.ResourceHandle()
+    >>> graph_props = pylibcugraph.experimental.GraphProperties(
+    ...     is_symmetric=False, is_multigraph=False)
+    >>> G = pylibcugraph.experimental.SGGraph(
+    ...     resource_handle, graph_props, srcs, dsts, weights,
+    ...     store_transposed=False, renumber=False, do_expensive_check=False)
+    >>> (paths, weights, sizes) = pylibcugraph.experimental.node2vec(
+    ...                             resource_handle, G, seeds, 3, True, 1.0, 1.0)
+
+    """
+
+    # FIXME: import these modules here for now until a better pattern can be
+    # used for optional imports (perhaps 'import_optional()' from cugraph), or
+    # these are made hard dependencies.
+    try:
+        import cupy
+    except ModuleNotFoundError:
+        raise RuntimeError("node2vec requires the cupy package, which could not "
+                           "be imported")
+    try:
+        import numpy
+    except ModuleNotFoundError:
+        raise RuntimeError("node2vec requires the numpy package, which could not "
+                           "be imported")
+    assert_CAI_type(src_array, "src_array")
+
+    cdef cugraph_resource_handle_t* c_resource_handle_ptr = \
+        resource_handle.c_resource_handle_ptr
+    cdef cugraph_graph_t* c_graph_ptr = graph.c_graph_ptr
+
+    cdef cugraph_random_walk_result_t* result_ptr
+    cdef cugraph_error_code_t error_code
+    cdef cugraph_error_t* error_ptr
+
+    cdef uintptr_t cai_srcs_ptr = \
+        src_array.__cuda_array_interface__["data"][0]
+    cdef cugraph_type_erased_device_array_view_t* srcs_view_ptr = \
+        cugraph_type_erased_device_array_view_create(
+            <void*>cai_srcs_ptr,
+            len(src_array),
+            get_c_type_from_numpy_type(src_array.dtype))
+
+
+    error_code = cugraph_node2vec(c_resource_handle_ptr,
+                                  c_graph_ptr,
+                                  srcs_view_ptr,
+                                  max_depth,
+                                  compress_result,
+                                  p,
+                                  q,
+                                  &result_ptr,
+                                  &error_ptr)
+    assert_success(error_code, error_ptr, "cugraph_node2vec")
+
+    # Extract individual device array pointers from result and copy to cupy
+    # arrays for returning.
+    cdef cugraph_type_erased_device_array_view_t* paths_ptr = \
+        cugraph_random_walk_result_get_paths(result_ptr)
+    cdef cugraph_type_erased_device_array_view_t* weights_ptr = \
+        cugraph_random_walk_result_get_weights(result_ptr)
+    cdef cugraph_type_erased_device_array_view_t* path_sizes_ptr = \
+        cugraph_random_walk_result_get_path_sizes(result_ptr)
+
+    cupy_paths = copy_to_cupy_array(c_resource_handle_ptr, paths_ptr)
```

Comment:
I liked how the function `copy_to_cupy_array` eased the code from the previous implementation